cmd2/AutoCompleter.py

```diff
@@ -0,0 +1,796 @@
+        # if we don't have a flag to populate with arguments and the last token starts with
+        # a flag prefix then we'll complete the list of flag options
+        completion_results = []
+        if not flag_arg.needed and len(tokens[-1]) > 0 and tokens[-1][0] in self._parser.prefix_chars:
+            return AutoCompleter.basic_complete(text, line, begidx, endidx,
+                                                [flag for flag in self._flags if flag not in matched_flags])
+        # we're not at a positional argument, see if we're in a flag argument
+        elif not current_is_positional:
+            # current_items = []
+            if flag_action is not None:
+                consumed = consumed_arg_values[flag_action.dest]\
+                    if flag_action.dest in consumed_arg_values.keys() else []
+                # current_items.extend(self._resolve_choices_for_arg(flag_action, consumed))
+                completion_results = self._complete_for_arg(flag_action, text, line, begidx, endidx, consumed)
+                if not completion_results:
+                    self._print_action_help(flag_action)
+
+        # ok, we're not a flag, see if there's a positional argument to complete
+        else:
+            if pos_action is not None:
+                pos_name = pos_action.dest
+                consumed = consumed_arg_values[pos_name] if pos_name in consumed_arg_values.keys() else []
+                completion_results = self._complete_for_arg(pos_action, text, line, begidx, endidx, consumed)
+                if not completion_results:
+                    self._print_action_help(pos_action)
+
+        return completion_results
+
+    @staticmethod
+    def _process_action_nargs(action: argparse.Action, arg_state: _ArgumentState) -> None:
+        if isinstance(action, _RangeAction):
+            arg_state.min = action.nargs_min
+            arg_state.max = action.nargs_max
+            arg_state.variable = True
+        if arg_state.min is None or arg_state.max is None:
+            if action.nargs is None:
+                arg_state.min = 1
+                arg_state.max = 1
+            elif action.nargs == '+':
+                arg_state.min = 1
+                arg_state.max = float('inf')
+                arg_state.variable = True
+            elif action.nargs == '*':
+                arg_state.min = 0
+                arg_state.max = float('inf')
+                arg_state.variable = True
+            elif action.nargs == '?':
+                arg_state.min = 0
+                arg_state.max = 1
+                arg_state.variable = True
+            else:
+                arg_state.min = action.nargs
+                arg_state.max = action.nargs
+
+    def _complete_for_arg(self, action: argparse.Action,
+                          text: str,
+                          line: str,
+                          begidx: int,
+                          endidx: int,
+                          used_values=list()) -> List[str]:
```

Comment:
**WARNING: Mutable default arguments in Python behave like static local variables in C** - they are forever persistent and if you change the value within the function, you have permanently altered the default for the next time you call the function.

If this is how you intend this function to work, then please document to make it clear.

If this is NOT how you intend this function to work, then use a non-mutable default argument.  If a tuple can work for you then that is fine.  If you really need that argument to be a list then you can default it to ``None`` and then immediately first thing within the function check if it is None and if so assign an empty list to it.